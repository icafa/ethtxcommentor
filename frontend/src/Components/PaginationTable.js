import React, { useState, useEffect } from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { Text, Table, Button } from '@aragon/ui'

import { GU } from '../utils/utils'

const PaginationWrapper = styled.div`
  display: flex;
  justify-content: space-between;
  width: 100%;
  padding: 0 ${GU / 2}px 0 ${GU / 2}px;
  margin-bottom: ${GU}px;
`

const PER_PAGE = 10

const PaginationTable = ({ header, items, children }) => {
  const [paginatedItems, setPaginatedItems] = useState([])
  const [paginationStart, setPaginationStart] = useState(1)
  const [paginationEnd, setPaginationEnd] = useState(10)
  const [currentPage, setCurrentPage] = useState(1)

  // Pagination logic
  useEffect(() => {
    // if there are more than 10 transactions, calculate the corresponding position, otherwise, assign the array length
    let paginationEnd =
      items.length > 10 ? currentPage * PER_PAGE : items.length
    // if paginationEnd is bigger than the items array, assign the array length
    if (paginationEnd > items.length) {
      paginationEnd = items.length
    }
    // if there are ten items or less, pagination start will always be one
    let paginationStart = items.length > 10 ? paginationEnd - PER_PAGE + 1 : 1
    // if we are at the end of the array, we only wanna show the remaining items, not necessarily 10
    if (paginationEnd === items.length) {
      const remaining = paginationEnd % 10
      // if remaining is not at 0, there are items left
      if (remaining !== 0) {
        paginationStart = paginationEnd - remaining
      }
    }
    // but if pagination end is less than then, it means we must be at one.
    if (paginationEnd < 10) {
      paginationStart = 1
    }
    setPaginationStart(paginationStart)
    setPaginationEnd(paginationEnd)
  }, [currentPage, items.length])

  // pick the corresponding items for the array
  useEffect(() => {
    setPaginatedItems(items.slice(paginationStart - 1, paginationEnd))
  }, [items, paginationEnd, paginationStart])

  return (
    <>
      {items.length === 0 && (
        <div>
          <Text smallcaps>No items to show.</Text>
        </div>
      )}
      {items.length > 0 && (
        <>
          <PaginationWrapper>
            <Text smallcaps data-testid="pagination-text">
              {paginationStart} - {paginationEnd} out of {items.length}
            </Text>
            <div>
              <Button
                css={`
                  margin-right: ${GU / 2}px;
                `}
                onClick={() => setCurrentPage(1)}
                disabled={currentPage === 1}
                data-testid="start-button"
              >
                <u>&lt;</u>
              </Button>
              <Button
                css={`
                  margin-right: ${GU / 2}px;
                `}
                onClick={() => setCurrentPage(page => page - 1)}
                disabled={currentPage === 1}
                data-testid="back-button"
              >
                &lt;
              </Button>
              <Button
                onClick={() => setCurrentPage(page => page + 1)}
                disabled={paginationEnd === items.length}
                data-testid="next-button"
              >
                &gt;
              </Button>
              <Button
                onClick={() =>
                  setCurrentPage(() => {
                    let lastPage = Math.floor(items.length / PER_PAGE)
                    if (items.length % PER_PAGE !== 0) {
                      lastPage++
                    }
                    return lastPage
                  })
                }
                disabled={paginationEnd === items.length}
                data-testid="end-button"
              >
                <u> &gt;</u>
              </Button>
            </div>
          </PaginationWrapper>

          <Table header={header}>{children(paginatedItems)}</Table>
        </>
      )}
    </>
  )
}

PaginationTable.propTypes = {
  header: PropTypes.node.isRequired,
  items: PropTypes.array.isRequired,
  children: PropTypes.func.isRequired,
}

export default PaginationTable
