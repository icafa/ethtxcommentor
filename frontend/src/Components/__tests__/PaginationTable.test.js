import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import { Main, TableRow, TableHeader, TableCell } from '@aragon/ui'
import PaginationTable from '../PaginationTable'

const mockedItems = [
  {
    thing: 'Guitar',
    strings: 6,
  },
  {
    thing: 'Bass',
    strings: 4,
  },
]

const paginableMockedItems = Array(20).map((value, index) => ({ value: index }))
const unevenPaginableMockedItems = Array(23).map((value, index) => ({
  value: index,
}))
describe('PaginationTable Component', () => {
  it('renders correctly with items', () => {
    const { asFragment } = render(
      <Main>
        <PaginationTable
          header={
            <TableRow>
              <TableHeader>Thing</TableHeader>
              <TableHeader>Strings</TableHeader>
            </TableRow>
          }
          items={mockedItems}
        >
          {items =>
            items.map(item => (
              <TableRow key={item.thing}>
                <TableCell>{item.thing}</TableCell>
                <TableCell>{item.strings}</TableCell>
              </TableRow>
            ))
          }
        </PaginationTable>
      </Main>
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it('paginates correctly', () => {
    const { getByTestId } = render(
      <Main>
        <PaginationTable
          header={
            <TableRow>
              <TableHeader>Value</TableHeader>
            </TableRow>
          }
          items={paginableMockedItems}
        >
          {items =>
            items.map(item => (
              <TableRow>
                <TableCell>{items.value}</TableCell>
              </TableRow>
            ))
          }
        </PaginationTable>
      </Main>
    )
    const paginationText = getByTestId('pagination-text')
    expect(paginationText).toHaveTextContent('1 - 10 out of 20')
    const backButton = getByTestId('back-button')
    expect(backButton).toBeDisabled()
    const startButton = getByTestId('start-button')
    expect(startButton).toBeDisabled()
    const nextButton = getByTestId('next-button')
    expect(nextButton).toBeInTheDocument()
    fireEvent.click(nextButton)
    expect(paginationText).toHaveTextContent('11 - 20 out of 20')
    fireEvent.click(nextButton)
    expect(paginationText).toHaveTextContent('11 - 20 out of 20')
    expect(nextButton).toBeDisabled()
    const endButton = getByTestId('end-button')
    expect(endButton).toBeDisabled()
    fireEvent.click(backButton)
    expect(paginationText).toHaveTextContent('1 - 10 out of 20')
    expect(startButton).toBeDisabled()
  })
  it('paginates correctly with a non-round array length', () => {
    const { getByTestId } = render(
      <Main>
        <PaginationTable
          header={
            <TableRow>
              <TableHeader>Value</TableHeader>
            </TableRow>
          }
          items={unevenPaginableMockedItems}
        >
          {items =>
            items.map(item => (
              <TableRow>
                <TableCell>{items.value}</TableCell>
              </TableRow>
            ))
          }
        </PaginationTable>
      </Main>
    )
    const paginationText = getByTestId('pagination-text')
    expect(paginationText).toHaveTextContent('1 - 10 out of 23')
    const backButton = getByTestId('back-button')
    expect(backButton).toBeDisabled()
    const startButton = getByTestId('start-button')
    expect(startButton).toBeDisabled()
    const nextButton = getByTestId('next-button')
    expect(nextButton).toBeInTheDocument()
    fireEvent.click(nextButton)
    expect(paginationText).toHaveTextContent('11 - 20 out of 23')
    fireEvent.click(nextButton)
    expect(paginationText).toHaveTextContent('20 - 23 out of 23')
    expect(nextButton).toBeDisabled()
    const endButton = getByTestId('end-button')
    expect(endButton).toBeDisabled()
    fireEvent.click(backButton)
    expect(paginationText).toHaveTextContent('11 - 20 out of 23')
    fireEvent.click(startButton)
    expect(startButton).toBeDisabled()
    expect(paginationText).toHaveTextContent('1 - 10 out of 23')
  })
})
