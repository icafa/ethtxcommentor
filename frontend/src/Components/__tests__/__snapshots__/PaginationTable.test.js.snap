// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationTable Component renders correctly with items 1`] = `
<DocumentFragment>
  <div>
    <div
      class="sc-cHGsZl OpiEb"
    >
      <span
        class="Text-yxw4o9-0 Hwozf"
        data-testid="pagination-text"
      >
        1 - 2 out of 2
      </span>
      <div>
        <button
          class="Button__StyledButton-sc-8npd5h-0 hTPQnb"
          data-testid="start-button"
          disabled=""
          type="button"
        >
          <u>
            &lt;
          </u>
        </button>
        <button
          class="Button__StyledButton-sc-8npd5h-0 hTPQnb"
          data-testid="back-button"
          disabled=""
          type="button"
        >
          &lt;
        </button>
        <button
          class="Button__StyledButton-sc-8npd5h-0 hTPQnb"
          data-testid="next-button"
          disabled=""
          type="button"
        >
          &gt;
        </button>
        <button
          class="Button__StyledButton-sc-8npd5h-0 hTPQnb"
          data-testid="end-button"
          disabled=""
          type="button"
        >
          <u>
             &gt;
          </u>
        </button>
      </div>
    </div>
    <table
      class="Table__StyledTable-uvcan9-0 cRowBj"
    >
      <thead>
        <tr
          class="TableRow__StyledTableRow-sc-3jn6yz-0 cFtmyC"
        >
          <th
            class="TableHeader__StyledTableHeader-sc-1qxm8cp-0 hKOaZg"
          >
            <div
              class="Text-yxw4o9-0 bBnSPr"
              color="#707070"
            />
          </th>
          <th
            class="TableHeader__StyledTableHeader-sc-1qxm8cp-0 hKOaZg"
          >
            <div
              class="Text-yxw4o9-0 bBnSPr"
              color="#707070"
            />
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="TableRow__StyledTableRow-sc-3jn6yz-0 cFtmyC"
        >
          <td
            class="TableCell__StyledTableCell-sc-110j155-0 hqZwhV"
          >
            <div
              class="TableCell__StyledTableCellContent-sc-110j155-1 klobR"
            >
              Guitar
            </div>
          </td>
          <td
            class="TableCell__StyledTableCell-sc-110j155-0 hqZwhV"
          >
            <div
              class="TableCell__StyledTableCellContent-sc-110j155-1 klobR"
            >
              6
            </div>
          </td>
        </tr>
        <tr
          class="TableRow__StyledTableRow-sc-3jn6yz-0 cFtmyC"
        >
          <td
            class="TableCell__StyledTableCell-sc-110j155-0 hqZwhV"
          >
            <div
              class="TableCell__StyledTableCellContent-sc-110j155-1 klobR"
            >
              Bass
            </div>
          </td>
          <td
            class="TableCell__StyledTableCell-sc-110j155-0 hqZwhV"
          >
            <div
              class="TableCell__StyledTableCellContent-sc-110j155-1 klobR"
            >
              4
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="ToastHub__Container-sc-1y0i8xl-0 stXpK"
    />
  </div>
</DocumentFragment>
`;
